lightning/src/debug_sync.rs

```diff
@@ -0,0 +1,217 @@
+	}
+
+	pub fn notify_all(&self) { self.inner.notify_all(); }
+}
+
+thread_local! {
+	/// We track the set of locks currently held by a reference to their `MutexMetadata`
+	static MUTEXES_HELD: RefCell<HashSet<Arc<MutexMetadata>>> = RefCell::new(HashSet::new());
+}
+static MUTEX_IDX: AtomicUsize = AtomicUsize::new(0);
+
+/// Metadata about a single mutex, by id, the set of things locked-before it, and the backtrace of
+/// when the Mutex itself was constructed.
+struct MutexMetadata {
+	mutex_idx: u64,
+	locked_before: StdMutex<HashSet<Arc<MutexMetadata>>>,
+	#[cfg(feature = "backtrace")]
+	mutex_construction_bt: Backtrace,
+}
+impl PartialEq for MutexMetadata {
+	fn eq(&self, o: &MutexMetadata) -> bool { self.mutex_idx == o.mutex_idx }
+}
+impl Eq for MutexMetadata {}
+impl std::hash::Hash for MutexMetadata {
+	fn hash<H: std::hash::Hasher>(&self, hasher: &mut H) { hasher.write_u64(self.mutex_idx); }
+}
+
+pub struct Mutex<T: Sized> {
+	inner: StdMutex<T>,
+	deps: Arc<MutexMetadata>,
+	skip_checks: bool
+}
+
+#[must_use = "if unused the Mutex will immediately unlock"]
+pub struct MutexGuard<'a, T: Sized + 'a> {
+	mutex: &'a Mutex<T>,
+	lock: StdMutexGuard<'a, T>,
+}
+
+impl<'a, T: Sized> MutexGuard<'a, T> {
+	fn into_inner(self) -> StdMutexGuard<'a, T> {
+		// Somewhat unclear why we cannot move out of self.lock, but doing so gets E0509.
+		unsafe {
+			let v: StdMutexGuard<'a, T> = std::ptr::read(&self.lock);
+			std::mem::forget(self);
+			v
+		}
+	}
+}
+
+impl<T: Sized> Drop for MutexGuard<'_, T> {
+	fn drop(&mut self) {
+		MUTEXES_HELD.with(|held| {
+			held.borrow_mut().remove(&self.mutex.deps);
+		});
+	}
+}
+
+impl<T: Sized> Deref for MutexGuard<'_, T> {
+	type Target = T;
+
+	fn deref(&self) -> &T {
+		&self.lock.deref()
+	}
+}
+
+impl<T: Sized> DerefMut for MutexGuard<'_, T> {
+	fn deref_mut(&mut self) -> &mut T {
+		self.lock.deref_mut()
+	}
+}
+
+impl<T> Mutex<T> {
+	pub fn new(inner: T) -> Mutex<T> {
+		Mutex {
+			inner: StdMutex::new(inner),
+			deps: Arc::new(MutexMetadata {
+				locked_before: StdMutex::new(HashSet::new()),
+				mutex_idx: MUTEX_IDX.fetch_add(1, Ordering::Relaxed) as u64,
+				#[cfg(feature = "backtrace")]
+				mutex_construction_bt: Backtrace::new(),
+			}),
+			skip_checks: false,
+		}
+	}
+
+	pub fn lock<'a>(&'a self) -> LockResult<MutexGuard<'a, T>> {
+		if !self.skip_checks {
+			MUTEXES_HELD.with(|held| {
+				// For each mutex which is currently locked, check that no mutex's locked-before
+				// set includes the mutex we're about to lock, which would imply a lockorder
+				// inversion.
+				for locked in held.borrow().iter() {
+					for locked_dep in locked.locked_before.lock().unwrap().iter() {
+						if *locked_dep == self.deps {
+							#[cfg(feature = "backtrace")]
+							panic!("Tried to violate existing lockorder. Mutex that should be locked after the current lock was created at: {:?}", locked.mutex_construction_bt);
```

Comment:
The backtrace was a bit confusing to me when testing this out since you need to set `RUST_BACKTRACE=1` to see the the offending backtrace. May want to indicate that in the message about using the backtrace feature.